Add simple updating part 2.

makson96 opened the pull request:
Hi,
as this pull request:
https://github.com/sharkwouter/minigalaxy/pull/152
seems to be staled for months I decided to pick it up from where it was left.
I have updated this branch to current master branch and removed error from your last comment.

Could you please make a review of what else problems need to be solved?
Are issues mentioned by you in last pull request still relevant (it does seem to work quick for me)?
Are some changes outdated and need to be removed or refactored?

I hope that I will be able to finally address this critical missing feature.
Best regards,
Tomek

sharkwouter commented:
Thanks for working on this. The old pull request died for a couple of reasons which you should know:

- I did not have time to review it well at the time. This time around I'll have more time, so this shouldn't be a problem.
- The PR was too big and contained more than one feature. This made it much harder for me to review it.
- Every time I tested the PR I found some bugs which would block the PR from being merged. The most annoying one is that games which didn't have updates available would show they do in some cases.

So this PR would need to include the following changes:

- It should only add update support. Other features should get their own PRs.
- It should not contain any merge conflicts.
- It should work and not introduce any new major bugs.

Hopefully this helps you to get this to work. This feature has been on the backlog for so long and the PR was just too big for me to do much with it initially. I really appreciate you want to work on this.

One more thing to keep in mind with this, is that I'm planning to finish and merge https://github.com/sharkwouter/minigalaxy/pull/176 this weekend. This will cause some conflicts and it is likely it will be merged before this PR. Sorry for that.

makson96 commented:
I will call it a day.
I have removed tons of code not directly related to update process. Now this pull request is less than 200 lines of changes so it should be easy for review.
I have also removed merge conflict. Don't worry about introducing another conflicts. I can handle that.

Thanks for quick response. I really appreciate it and I hope that I can be of use for this project.
Please also note, that:
- update process is broken at the moment. I will fix it probably tomorrow.
- there seems to be still some update unrelated code here. I will clean it more probably tomorrow.

Best regards and have a nice day.

sharkwouter commented:
Thank you very much for the work. Let me know when it is ready for review. I'll make some time for it on the weekend.

makson96 commented:
Hi,
I am ready for the review. Few notes.

- I have tested it only for Linux.
- You can test update procedure by going to game directory, opening "gameinfo" and changing second line of the file (this will make update available
- I have removed "UPDATE_QUEUED" and "UPDATE_DOWNLOADING" and use general "QUEUED" and "DOWNLOADING" instead. If you prefer custom message in case of downloading updates I can bring them back.
- new function "__move_and_overwrite" is designed to be safe for additional files already present in game directory (save games, configs, etc.)

sharkwouter commented:
![store-page](https://user-images.githubusercontent.com/5042659/96363746-5d878200-1136-11eb-95a2-d3c62a1f50f7.png)
Here is an example of where you can find the store page link. It only applies to installed games, though.

sharkwouter commented:
Unfortunately, when opening minigalaxy with this code my library is not fully loaded, only FTL is shown. In the command line I'm seeing the following exception:
```
wouter@wouter-pc:~/Sources/minigalaxy$ bin/minigalaxy 
Traceback (most recent call last):
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/library.py", line 108, in __create_gametiles
    self.__add_gametile(game)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/library.py", line 111, in __add_gametile
    self.flowbox.add(GameTile(self, game, self.api))
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/gametile.py", line 74, in __init__
    self.reload_state()
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/gametile.py", line 337, in reload_state
    installer = self.api.get_info(self.game)["downloads"]["installers"]
  File "/home/wouter/Sources/minigalaxy/minigalaxy/api.py", line 123, in get_info
    response = self.__request(request_url)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/api.py", line 190, in __request
    return response.json()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib/python3/dist-packages/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib/python3/dist-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
Fallout Classic (1) has no store page url
Fallout 2 Classic (2) has no store page url
Fallout Tactics Classic (3) has no store page url
Blood: One Unit Whole Blood (1207658856) has no store page url
Warsow (1207659121) has no store page url
Softporn Adventure (1207662103) has no store page url
Sudeki (1207664353) has no store page url
Duke Nukem (1207664963) has no store page url
Duke Nukem 2 (1207664973) has no store page url
Crystal Caves (1207665273) has no store page url
Cyberpunk 2077 Goodies Collection (1486144755) has no store page url
Traceback (most recent call last):
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/library.py", line 108, in __create_gametiles
    self.__add_gametile(game)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/library.py", line 111, in __add_gametile
    self.flowbox.add(GameTile(self, game, self.api))
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/gametile.py", line 74, in __init__
    self.reload_state()
  File "/home/wouter/Sources/minigalaxy/minigalaxy/ui/gametile.py", line 337, in reload_state
    installer = self.api.get_info(self.game)["downloads"]["installers"]
  File "/home/wouter/Sources/minigalaxy/minigalaxy/api.py", line 123, in get_info
    response = self.__request(request_url)
  File "/home/wouter/Sources/minigalaxy/minigalaxy/api.py", line 190, in __request
    return response.json()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib/python3/dist-packages/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib/python3/dist-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

I'll do a little bit of troubleshooting to see if I can find the issue.

sharkwouter commented:
You can start Minigalaxy with ``MG_DEBUG=1 bin/minigalaxy`` then you can see the API requests. I'm seeing this request causing an error:
```
Request: https://api.gog.com/products/0?expand=downloads,expanded_dlcs,description,screenshots,videos,related_products,changelog 
Return code: 404
```
The response body is an html page, not json.

sharkwouter commented:
It seems it thinks the id for the game PixelJunk Shooter is 0. Not sure what changed which could cause that. On to more debugging

sharkwouter commented on minigalaxy/ui/gametile.py:
```diff
@@ -272,10 +329,18 @@ def __get_install_dir(self):
         if self.current_state in dont_act_in_states:
             return
-        if self.game.install_dir and os.path.exists(self.game.install_dir):
+        if self.game.installed_version:
```
When ``self.game.id`` is set to 0, this shouldn't be executed. I'm not sure if it would execute again after the id is loaded from the API, though.

sharkwouter replied:
This happens with Pixeljunk Shooter. The gameinfo file for that game looks like this:
```
PixelJunk Shooter
2.0.0.3
r971
```

sharkwouter commented on minigalaxy/ui/gametile.py:
```diff
@@ -272,10 +329,18 @@ def __get_install_dir(self):
             return
-        if self.game.install_dir and os.path.exists(self.game.install_dir):
+        if self.game.installed_version:
+            installer = self.api.get_info(self.game)["downloads"]["installers"]
```
This line and the line below it have to be executed in a separate thread. It is currently making my Minigalaxy freeze up for a couple of seconds, because I have over 90 games installed.

makson96 commented:
Please check if change in "minigalaxy/ui/library.py" is responsible for this issue.

sharkwouter reviewed (changes requested):
Thanks for the PR. It seems to work with a small code change, but I'm still finding some other issues with it which will require a fix:

- After successfully updating a game, the update option remains on the gametile.
- If you make Minigalaxy only show installed games, it will no longer show which games can be updated.

One thing I didn't test was running Minigalaxy in offline mode with this change. That should still work.

it is really cool to see it mostly work, though, great work.

makson96 commented:
Hi,
Thank you for the great review. I believe that I have addressed all the issues mentioned by you. Please check this branch again.
Best regards,
Tomasz

sharkwouter commented:
Thanks! I'll try to take a look at this PR this evening, otherwise It will be on the weekend.

sharkwouter commented:
It seems to work well now and the code looks good. I'm impressed, thanks for the hard work!

sharkwouter commented:
I still want to say thank you to @mdgomes for the work he put into this as well. This change will be in the next release.

makson96 commented:
His original commits were included in this pull request, so his contribution can always be seen in the main project.
